google_analytics/aggregation.py

```diff
@@ -92,58 +93,193 @@ def aggregate(df):
         The same dataframe grouped and pivoted, where each visitor is a single row.
 
     """
-    df['transactions'].fillna(0, inplace=True)
-    df['transactionRevenue'].fillna(0.0, inplace=True)
+    # Long to wide pivot - one row per visitor.
+    dynamic_columns = ['transactionRevenue', 'visits', 'transactions', 'pageviews', 'Avg. Session Duration', 'Bounce Rate', ]
+    wide = pd.pivot_table(df, index='fullVisitorId', columns='date_temp', values=dynamic_columns)
 
-    def agg_date(s):
-        date = pd.datetime.strptime(s, '%Y-%m-%d')
-        return "{}_{}".format(date.month, date.year)
+    # Collapse multi-index.
+    wide.columns = wide.columns.to_series().str.join('_')
+    wide.reset_index(inplace=True)
 
-    df["date"] = df["date"].apply(agg_date)
+    # Now let's also get the static columns.
+    static_columns = list(set(df.columns) - set(dynamic_columns) - {"date"} - {"date_temp"})
+    static = df[static_columns]
+
+    # Regroup on visitor.
+    print("Regroup again to get the static fields - this will take a while")
+    static_grouped = static.groupby("fullVisitorId", as_index=False)[static_columns].sum()
+    # Merge dynamic and static features.
+    df = static_grouped.merge(wide, on="fullVisitorId", how="inner")
+    return df[natsorted(df.columns, key=lambda y: y.lower())]
 
+
+def group_data(df):
+    """ Aggregates the data per user and date
+    """
     def most_frequent(x):
         return x.value_counts().index[0]
 
     agg = {
-        "operatingSystem": most_frequent,
-        "country": most_frequent,
-        "browser": most_frequent,
-        "pageviews": sum,
-        "transactions": sum,
-        "visits": sum,
-        "transactionRevenue": sum,
-        "visitStartTime": "first"
-    }
+            "operatingSystem": most_frequent,
+            "country": most_frequent,
+            "browser": most_frequent,
+            "weekday": most_frequent,
+            "pageviews": sum,
+            "transactions": sum,
+            "visits": sum,
+            "Avg. Session Duration": sum,
```

Comment:
This needs to be aggregated by the average

Reply:
+ 1, same below